src/main/java/systems/composable/dropwizard/cassandra/CassandraProvider.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2017 Composable Systems Limited
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package systems.composable.dropwizard.cassandra;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Session;
+import io.dropwizard.setup.Environment;
+import org.glassfish.hk2.api.Factory;
+import org.glassfish.hk2.utilities.binding.AbstractBinder;
+import org.glassfish.jersey.process.internal.RequestScoped;
+
+import javax.inject.Singleton;
+
+/**
+ * Provides logic for registering factories that provide and dispose injected instances for
+ * {@link javax.ws.rs.core.Context}-annotated
+ * {@link Cluster} and {@link Session} classes.
+ *
+ * @author <a href="mailto:max@dominichenko.com">Max Dominichenko</a>
+ * @see CassandraBundle
+ */
+class CassandraProvider {
+
+	/**
+	 * Encapsulates logic that binds {@link Cluster} and {@link Session} classes to
+	 * {@link Singleton} or {@link RequestScoped} interfaces respectively.
+	 */
+	public static class Binder extends AbstractBinder {
+
+		private final CassandraProvider cassandraProvider;
+
+		Binder(CassandraFactory cassandraFactory, Environment environment) {
+			cassandraProvider = new CassandraProvider(cassandraFactory, environment);
+		}
+
+		@Override
+		protected void configure() {
+			bindFactory(cassandraProvider.clusterFactory).to(Cluster.class).in(Singleton.class);
+			bindFactory(cassandraProvider.sessionFactory).to(Session.class).in(RequestScoped.class);
```

Comment:
If I'm reading this correctly, a new `Session` is instantiated for each request - is that right?

The docs suggest using a single `Session` per application (see http://docs.datastax.com/en/drivers/java/3.1/com/datastax/driver/core/Session.html)

Past versions of the driver had a memory leak when closing a `Session`, which lead to a steady rise in memory usage (closed sessions weren't being GC'd). I'm not sure if this has been fixed, but I think it would be better to make the `Session` a singleton.